cmd/lambda/leases/get.go

```diff
@@ -1,39 +1,139 @@
+		response.WriteNotFoundError(w)
+		return
+	}
+
+	leaseResponse := response.LeaseResponse(*lease)
+	json.NewEncoder(w).Encode(leaseResponse)
 }
 
-// Call - function to return a specific AWS Lease record to the request
-func (controller GetController) Call(ctx context.Context, req *events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
+// GetLeasesByPrincipcalIDAndAccountID - Returns a list of leases by principal and account
+func GetLeasesByPrincipcalIDAndAccountID(w http.ResponseWriter, r *http.Request) {
 	// Fetch the account.
-	leaseID := path.Base(req.Path)
-	lease, err := controller.Dao.GetLeaseByID(leaseID)
+	principalID := r.FormValue(PrincipalIDParam)
+	accountID := r.FormValue(AccountIDParam)
+	lease, err := dao.GetLease(accountID, principalID)
 	if err != nil {
-		log.Printf("Error Getting Lease for Id: %s", leaseID)
-		return response.CreateAPIErrorResponse(http.StatusInternalServerError,
-			response.CreateErrorResponse("ServerError",
-				fmt.Sprintf("Failed Get on Lease %s",
-					leaseID))), nil
+		errMsg := fmt.Sprintf("Error getting lease for principal %s and acccount %s: %s", principalID, accountID, err.Error())
+		log.Println(errMsg)
+		response.WriteServerErrorWithResponse(w, errMsg)
+		return
 	}
 	if lease == nil {
 		log.Printf("Error Getting Lease for Id: %s", err)
-		return response.NotFoundError(), nil
+		response.WriteNotFoundError(w)
+		return
 	}
 
 	leaseResponse := response.LeaseResponse(*lease)
-	return response.CreateJSONResponse(http.StatusOK, leaseResponse), nil
+	json.NewEncoder(w).Encode(leaseResponse)
+}
+
+// GetLeasesByPrincipalID - Returns a list of leases by principal and account
+func GetLeasesByPrincipalID(w http.ResponseWriter, r *http.Request) {
+	// Fetch the account.
+	principalID := r.FormValue(PrincipalIDParam)
+	leases, err := dao.FindLeasesByPrincipal(principalID)
+	if err != nil {
+		errMsg := fmt.Sprintf("Error getting leases for principal %s: %s", principalID, err.Error())
+		log.Println(errMsg)
+		response.WriteServerErrorWithResponse(w, errMsg)
+		return
+	}
+
+	if len(leases) == 0 {
+		response.WriteNotFoundError(w)
+		return
+	}
+
+	leaseResponses := []*response.LeaseResponse{}
+
+	for _, l := range leases {
+		leaseResponse := response.LeaseResponse(*l)
+		leaseResponses = append(leaseResponses, &leaseResponse)
+	}
+
+	json.NewEncoder(w).Encode(leaseResponses)
+}
+
+// GetLeasesByAccountID - Returns a list of leases by principal and account
+func GetLeasesByAccountID(w http.ResponseWriter, r *http.Request) {
+	// Fetch the account.
+	accountID := r.FormValue(AccountIDParam)
+	leases, err := dao.FindLeasesByAccount(accountID)
+	if err != nil {
+		errMsg := fmt.Sprintf("Error getting leases for account %s: %s", accountID, err.Error())
+		log.Println(errMsg)
+		response.WriteServerErrorWithResponse(w, errMsg)
+		return
+	}
+
+	if len(leases) == 0 {
+		// We were throwing an error on these, but not sure that's the right thing
```

Comment:
Pardon me if I am over explaining here - given that it seems you've thought about this yourself via this comment, but a query with no results should not throw an HTTP error, but rather return a status 200 with an "empty" (like a JSON array without an elements) body. 404s are used for resources in path parameters that are not found. In the case of querying a collection with zero results, there really wasn't an error. If you query, nothing breaks, and you have a result empty result set, then you still have the correct results, which in this case happen to be empty.

Conclusion here being that this should be change to 200 OK with empty array as body.